Migration to Vite

sronveaux opened the pull request:
Hi,

I eventually found the time to work on one of the major steps in the `Roadmap for Wegue V3`, migration to `Vite` !

I think it's time to share a draft PR here which should be fully functional but still needs a few points to be addressed properly. For this, I need your input to do things in a way that would satisfy everybody...

Many things have changed between `Vue-CLI` and `Vite` so lots of breaking changes have once again arised but I do really hope that by following the migration guide you will be able to test things fast and smooth !

Don't hesitate to ask questions and precisions here though so I can explain the design choices I did in case you wonder...

You'll see that there are two warnings that keep firing each time the `Wegue` page is requested in development. Those will vanish as soon as the `compat build` of `Vue` will be removed and are thus not of concern. A built app doesn't produce them neither. 

Concerning the points where I need your input, here they are:
- There's also a warning that appears every time the unit tests are run. It has no impact on the tests themselves and seems harmless.  
Please note that this warning is also present with the minimal example given inside `karma-vite` repository. I asked the maitainer whether he knew where this comes from but haven't got an answer yet. It should be possible to silence this warning by changing the logLevel inside `vite.config.js` but I think we could miss some other important things so didn't do it. Just tell me if you agree with that.  
I'm also planning to give `Web Test Runner` a try to replace `Karma` completely as strongly suggested by `Karma` team. If this work, this will automatically be fixed then.
- `Vue-CLI` was using `npm run serve` and `npm run build` and was configured in `Wegue` to serve pages on port 8081 and open the browser automatically when `npm run serve` was run.  
Applications bootstrapped with `vue create` today use `npm run dev`, `npm run build` and `npm run preview`. Dev pages are served on port 5173 and build preview on port 4173 without opening the browser by default.  
I currently aligned with the current `vue create` skeleton but added the two lines needed to open the browser automatically when `npm run dev` or `npm run preview` are launched. Documentation was amended with this information. Are you O.K. with this ?
- When bootstrapping a new application using `vue create`, `ESLint` is installed and configured but rules are not enforced while building. In Vite, there's nothing automatic and you have to run the script `npm run lint` manually yourself.  
I saw there were some plugins available like [vite-plugin-eslint2](https://www.npmjs.com/package/vite-plugin-eslint2) or the [@nabla/vite-plugin-eslint](https://www.npmjs.com/package/@nabla/vite-plugin-eslint) which could partially put this feature back.  
According to [vite-plugin-eslint2 FAQ](https://vite-plugin-eslint2.modyqyw.top/guide/faq), this plugin is not needed most of the time as modern IDE already do the job for us and because it slows `Vite` down. But it seems I could give a try to `@nabla/vite-plugin-eslint` which seems to be faster but will never make a dev build fail as it was the case with `Vue-CLI` and would be possible with `vite-plugin-eslint2`. It will just report errors and warnings in the console.  
What do you think would be good or acceptable for Wegue V3 ?  Do you want me to try one of those plugins or do you also think this feature is not as useful today ?  
If this looks important for you, I'd be in favor of using `@nabla/vite-plugin-eslint` even if it has limitations so not to slow `Vite` down too much. I could also try to adapt the build script to run the linter first and only build if it succeeds (as now with `Vue-CLI`). Development would then always succeed, reporting errors and warnings in the console only but you would not be able to build the app if there are some linting erros as it is the case now. What do you think about this?
- Static files management works very differently in `Vite`. In current `Wegue` version, the contents of `app/static` are copied in `dist/static`. This is not as easy in `Vite` because you can specify a static directory but it's then copied in the root `dest` folder next to HTML templates and there's no easy way to change that. There is a `vite-plugin-static-copy` plugin that can be used but there's a risk because paths will not be translated at build time as it is the case for now. Moreover, when run in dev mode, it copies the files once and for all at startup so any change in a config file for example will imply to stop and relaunch the dev server which is a real problem from my point on view.  
The two simple solutions are to use default `Vite` behaviour and so have all static files inside the `dist` folder and not inside `dist/static` or to move the source files from `app/static` to `app/static/static`.  
In the first case, this implies to modify the configuration files so data are loaded from `/data` instead of `/static/data`. This is how I've currently adapted `Wegue` so you can compare the results inside the `dist` directory by yourself. The result is not as clean as before for sure but it's a good compromise from my point of view.  
In the second case it should work as now without changing anything but having this `app/static/static` folder is quite strange I think. We could rename the top `static` folder to something like `assets` so the contents of `app/assets/static` would end in `dist/static` as is currently the case.  
Just tell me your point of view on this point...
- Last point is a design point about the `app/static/css/app.css` file which is first created empty by `init:app` and often used to style an app. Historically, this file was not present in the `app-starter` directory so there was some code inside `main.js` which tried to load it as it was optional and potentially inexistent. This code relied on `Webpack require context` and was migrated using `Vite glob import` which in this case does the same, but `Vite` emits some warnings about this file for some good reasons I never thought about before. It is situated inside `app/static` and so it gets copied at build time but in fact it is already parsed and integrated during the build step. Modifying it afterwards doesn't apply dynamically. So it shouldn't be present in the `static` directory and shouldn't be copied while bundling.  
Moreover, as the file is now present inside the `app-starter`, trying to import the file as it is done now looks more like a fallback for legacy apps. I personally think this file should be imported as all other dependencies and moved somewhere else in the directory structure. But where to ? To stay coherent with `src` folder, should it be moved to a `app-starter/assets/css` or `app-starter/styles` directory ? That's what I'd personally suggest. Do you agree with that ?  
Eventually think moving it to `app-starter/assets/css` could clash with previous point if you think static files should also be moved to `app/assets/static`. That's the reason why I did nothing on those two last points and preferred waiting for your input!

I think that's all for now, I'll create some new comments if something else comes through my mind. Thanks in advance for your inputs on those four points and I really hope you'll enjoy the results !

Cheers,
Sébastien

sronveaux commented:
Hi @chrismayer,

So following @fschmenger comments, I think I'll go in that direction:

- Keep current configuration of ports and browser opening when running `npm run dev`, `npm run build` and `npm run preview`
- Rename the `app-starter/static` folder to `app-starter/public/static` so nothing else should be changed inside configuration files
- Move `.css` files to `app-starter/styles` and `src/styles` directories and import the custom app styles statically
- Give a try to [@nabla/vite-plugin-eslint](https://github.com/nabla/vite-plugin-eslint) and see if it's useful or not. I'm still not sure whether it will be used in the end or not

I'm still not sure whether `ESLint` should be enforced before running the build or not. What do you think about this specific point?  
In case you have different opinions about the other points, don't hesitate to write them here !

All the best,
Sébastien

fschmenger reviewed (commented):
Hi @sronveaux, I`m done with my review by now. I added 2 more little points for you to look into.

sronveaux commented on tests/unit/karma.conf.cjs:
Yes, that's exactly the "warning" I mentioned. This one appears even in the example project of `karma-vite`. You're right, it is an "error" which doesn't affect test functionality which is why I talked about a "warning".

I'm still waiting for an answer from the main developer about this one but have no answer yet.  
I suspect this is due to the fact that you **must** flag included `fileList` in `karma.conf.cjs` to `served: false` which tells `Karma` not to serve the file over the web server it automatically creates (otherwise nothing works at all) because `Vite` is in charge of it here, `Karma` is only there to launch the browser and trigger the tests.  
However, `Karma` certainly forget to remove some refenrence to it inside the the generated `HTML` file as we see the typical paths starting with `/base/` which is where `Karma` serves all the content it produces... so I suspect this is more a `Karma` issue than a `karma-vite` one...

sronveaux commented:
Hi @fschmenger, @chrismayer,

Here I updated the PR with all the comments I received. Here's a summary of what was done since the first publication of the PR:

- The `app-starter/static` folder was renamed to `app-starter/public/static`
- The `.css` files were moved to `app-starter/styles` and `src/styles` directories.
- The custom app CSS styles are now imported statically.
- `@nabla/vite-plugin-eslint` is included for now. As you'll see, it only emits warnings and errors inside the terminal where `Vite` is running. Nothing appears in the browser console and it has no other impact.
- ESLint is manually run before a build and needs to succeed for the build to take place.
- `<noscript>` blocks have been readded inside `HTML` templates.
- Language files import rules have been relaxed.
- Documentation and migration guide were amended accordingly.

So what is missing is comments inside the `vite.config.js`. These will be added when removing the draft status of the PR as some changes can still potentially take place.

Does this meets your expaectations like this ?

From here, I'd also like to have your return on ESLint configuration and plugin inclusion. Is it O.K. like this ?
And do you think adding `vite-overlay-plugin` mentioned in https://github.com/wegue-oss/wegue/pull/449#discussion_r2450925689 once its behaviour is stabilized could be interesting or not ?  If you think it is, I would be in favour of adding an issue but go forward with this PR without it. Do you agree ?

Thanks in advance for your return,

All the best,
Sébastien

sronveaux commented:
Hi @fschmenger and thanks for your return !

I'll be off next week so if someone has some ideas or comments in that time, don't hesitate to post !  
Otherwise, I'll try to add the comments and propose the PR officially beginning of November as soon as I'll have a little time...

sronveaux commented:
Hello,

I just added some commentaries to the `vite.config.js` file as they were still missing.

Concerning the open issue with `karma-vite`, still no answer from the developer... to be followed.

I also added a few lines inside the `.vscode/settings.json` file:

- This file was added in a2898166f00333e795bef7ae90b04941280d7c69 to conform to latest Vue skeleton template.
- In this default configuration, lint **AND** fix is done on each file save. As it was decided a long time ago to only report linting errors but not correct them automatically inside `npm` scripts, config was amended to follow the same behaviour if `ESLint` extension is used.
- The local `ESLint` version is explicitly used instead of a potential global one.
- `.vue` files linting was enabled as explained in [official eslint-plugin-vue documentation](https://eslint.vuejs.org/user-guide/#visual-studio-code).

This has no impact on the code itself, only if using `VSCode` and if official `ESLint` extension is installed which is one of the recommended extensions tagged, once again, in a2898166f00333e795bef7ae90b04941280d7c69.

As all topics have been addressed or discussed, I'll remove the draft state on the PR as it's ready for real review and potential merge.  
However, @fschmenger and @chrismayer, don't hesitate to ask more questions or request more changes if needed!

All the best,
Sébastien
